rust/src/data_structures.rs

```diff
@@ -84,7 +85,10 @@ impl ListenInfo {
                 Protocol::Udp => transport::Protocol::Udp,
             },
             ip: self.ip.to_string(),
-            announced_ip: self.announced_ip.clone().map(|ip| ip.to_string()),
+            announced_address: self
+                .announced_address
+                .clone()
+                .map(|address| address.to_string()),
```

Comment:
Two allocations look strange. I'm pretty sure this will also work:
```suggestion
            announced_address: self
                .announced_address
                .as_ref()
                .map(|address| address.to_string()),
```

Though this is not new, it was like this before, I just comment on what I noticed.

Reply:
You mean that this thing should also be done in other changes I did?

Reply:
Maybe, but I was commenting just about this one

Reply:
On it.

Reply:
Nope, I don't see any other place in which `as_ref()` can be used instead of cloning.